### PR TITLE
feat: refactor design docs

### PR DESCRIPTION
**Pull request by zw0610**
**What this PR does / why we need it**:

Refactor design doc

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @terrytangyuan @gaocegege @pendoragon 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: docs/design/communication_library.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+## Communication Library
+
+We believe there is no need to re-implement a communication library compatible with data structures differing from this dl framework to another. Therefore, FTLib will import a user-specific communication library during initialization.
```
Comment (terrytangyuan) [inline docs/design/communication_library.md]: data structures differing from this dl framework to another -> maybe "framework-specific data structures"

File: docs/design/ftlib.md [hunk 2]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
-1. New workers report joining in
-2. Collective-communication operations fail or time out
-3. *Liveness probe fails for existing workers*
 
-Here we assume the third case is or will eventually lead to the second one. Thusly, only the first and second occasions are handled explicitly in FTLib.
 
-*In the next design, we will consider hiding communication APIs of FTLib by wrapping model and optimizer, which hooks the communication operations for each parameter to its gradient calculation function. In this way, the optimizer will only need to check if all parameters hook are executed before updating weights.*
+## Passive or Active Mode
+
+With different consensus, FTLib wraps communication APIs with two modes:
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: consensus -> consensus methods

File: docs/design/communication_library.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+## Communication Library
+
+We believe there is no need to re-implement a communication library compatible with data structures differing from this dl framework to another. Therefore, FTLib will import a user-specific communication library during initialization. 
+
+Some deep learning frameworks wrap communication libraries, such like NCCL or Gloo, and make it compatible with its own data structure. Such wrapped communication libraries can be adopted with less effort to be compatible with deep learning frameworks.
+
+Since we are working with ElasticDL, some low-level APIS are exposed.
```
Comment (terrytangyuan) [inline docs/design/communication_library.md]: Maybe rephrase to "There are also deep learning frameworks that require highly customized usage of the collective communication APIs. In this case, we also expose several low-level APIs to facilitate that."?

File: docs/design/communication_library.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+Users are also able to combine operations from the communication library as well as the deep learning frameworks to customize a function and throw it to `FTLib.execute()` to perform all-in-one action.
```
Comment (terrytangyuan) [inline docs/design/communication_library.md]: Maybe add a sentence to explain what ``FTLib.execute()`` does and why would I use it instead of using combined operations directly

File: docs/design/communication_library.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+Given most communication libraries bound with dl frameworks require a initialization step, we ask the `Framework` to provide another two APIs: `rebuild` and `abort_communicator`. The `rebuild` API does the initialization stuff while the `abort_communicator` covers all de-initialization things.
```
Comment (terrytangyuan) [inline docs/design/communication_library.md]: dl -> deep learning
a -> an
Also maybe add a short sentence to give some background on (de-)initialization since readers may not know what's involved during those processes

File: docs/design/consensus.md [hunk 6]
```diff
@@ -1,24 +1,30 @@
 ## Consensus
 
-The consensus protocol in FTLib acts as a shadow precondition for any collective communication operations. Any changes from the consensus protocol will reset the initialization flag of FTLib to `False`, deterring any communication operations after the rebuild procedure returns success.
+The consensus protocol in FTLib acts as a shadow precondition for any collective communication operations. Any changes from the consensus protocol will reset the `_initialized` flag of FTLib to `False`, deterring any communication operations after the rebuild procedure returns success. Under passive or active mode, such change takes place immediately or eventually (when FTLib check the consensus next time).
```
Comment (terrytangyuan) [inline docs/design/consensus.md]: Under passive or active mode -> depending on the choice of passive or active mode

Comment (terrytangyuan) [inline docs/design/consensus.md]: Should we also expose `_initialized` to `initialized`?

File: docs/design/consensus.md [hunk 7]
```diff
@@ -1,24 +1,30 @@
...
 
-A member list is maintained by the implementation of consensus protocol.
+A member list is maintained by an implementation of consensus protocol.
 
-The rank-assign scheme in FTLib can extract worker identification from the member list, such like address of each worker. This unique identification helps the rank-assign scheme to designate individual rank number to each worker, which most communication libraries require when initializing.
+The rank-assign scheme in FTLib can extract worker identification from the member list, such like address of each worker, which helps the rank-assign scheme to designate individual rank number to each worker.
```
Comment (terrytangyuan) [inline docs/design/consensus.md]: Consider adding a short description on "rank-assign scheme" even though it's pretty intuitive to readers with background

File: docs/design/consensus.md [hunk 8]
```diff
@@ -1,24 +1,30 @@
...
 
-When FTLib start to `rebuild`, it uses the `confirm` API of consensus protocol to check the member list is agreed by all existing workers.
+As described in [`FTLib`](./ftlib.md), if active mode is on, the `confirm` API will be called for in `initialized()` and `skip_allreduce()`, consenting member list before communication API calls.
```
Comment (terrytangyuan) [inline docs/design/consensus.md]: Explain briefly what "consenting member list" implies

File: docs/design/consensus.md [hunk 9]
```diff
@@ -1,24 +1,30 @@
...
 
 Not exposed to FTLib though, a `report_join` API in consensus protocol will be called inside the `confirm` function if the worker is freshly launched and has not successfully reported before. During the whole lifetime of a worker, the `report_join` API will not be called for a second time after a successfully trial.
 
-Every time FTLib succeeds or fails to perform collective operations, it will call the corresponding functions to inform the consensus protocol whether any actions need to be taken. However, these two functions can be ignored if no actions needed by the specific consensus protocol.
+Every time FTLib succeeds or fails to perform collective operations, it will call the corresponding functions to inform the consensus whether any actions need to be taken. However, these two functions can be ignored if no actions needed by the specific consensus protocol.
```
Comment (terrytangyuan) [inline docs/design/consensus.md]: no actions needed -> no actions are needed

File: docs/design/ftlib.md [hunk 10]
```diff
@@ -6,10 +6,10 @@
 
 *Fig 1. General Architecture of FTLib*
 
-FTLib maintains a set of collective communication APIs for distributed training regardless worker loss of join. It consists of two major components:
+FTLib maintains a set of collective communication APIs as well as some high-level APIs for distributed training regardless worker loss of join. It consists of two major components:
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: collective communication APIs -> low-level collective communication APIs

File: docs/design/ftlib.md [hunk 11]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
 
 *Fig 2. The Role FTLib Played in Distributed Training*
 
-In the contemporary design, the FTLib is called explicitly to perform collective operation on all parameters of a model.
+In the contemporary design, the FTLib is called explicitly to perform collective operations or self-designed high-level calls.
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: What do you mean by "self-designed high-level calls"?

File: docs/design/ftlib.md [hunk 12]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
 
-For each batch of data, the model in each worker will be fed with training data, producing loss by forwarding and gradients by backward propagation. With local gradients ready on each worker, the allreduce API in FTLib will be called to average the gradients across all workers. As the operation is blocking, it indicates all parameters on this process is ready to be updated if allreduce API returns success. Otherwise, the training process will abort or continue according to different returned status.
+For each batch of data, the model in each worker will be fed with training data, producing loss by forwarding and gradients by backward propagation. With local gradients ready on each worker, the allreduce API in FTLib will be called to average the gradients across all workers. As the operation is blocking, it indicates all parameters on this process is ready to be updated if allreduce API returns success.
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: forwarding -> forward calculation

Comment (terrytangyuan) [inline docs/design/ftlib.md]: With local gradients ready on each worker -> Once local gradients are successfully calculated on each worker

File: docs/design/ftlib.md [hunk 13]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
 
-Before launching any communication operations, the FTLib will lock the process and unlock after finished or aborted. This design prevents the initialized flag from being modified by `Consensus.new_member_join` on other thread during the average phase.
+Otherwise, the training process will abort or continue according to different returned status.
 
-The initialized flag of a FTLib will be reset (to `False`) under three occasions:
+The entire step described above can be wrapped into a high-level customized API with FTLib, which hides the latency of communication by overlapping computation and communication.
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: I think this is what you meant by "self-designed high-level calls" that you mentioned earlier. Maybe explain this benefit earlier so readers would have more context

File: docs/design/ftlib.md [hunk 14]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
 
-Here we assume the third case is or will eventually lead to the second one. Thusly, only the first and second occasions are handled explicitly in FTLib.
 
-*In the next design, we will consider hiding communication APIs of FTLib by wrapping model and optimizer, which hooks the communication operations for each parameter to its gradient calculation function. In this way, the optimizer will only need to check if all parameters hook are executed before updating weights.*
+## Passive or Active Mode
+
+With different consensus, FTLib wraps communication APIs with two modes:
+
+- Active Mode: check member list consensus in `initialized()` and `skip_allreduce()` APIs  
+- Passive Mode: not perform the check described in activte mode
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: activte -> activate

File: docs/design/ftlib.md [hunk 15]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
+
+Before launching any communication operations, the FTLib will lock the process and unlock after finished or aborted. This design prevents the initialized flag from being modified by `Consensus.new_member_join`.
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: the process -> all processes in all workers

File: docs/design/ftlib.md [hunk 16]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
+## FTLib Flow with Low-Level APIs
 
-The flow chart below describes how FTLib works internally.
+The flow chart below describes how FTLib works internally when low-level APIs are used in distributed training.
 
 ![ftlib_flow](../imgs/ftlib_flow.png)
 
 *Fig 3. The Internal Mechanism of FTLib*
 
+The initialized flag of a FTLib will be reset (to `False`) under three occasions:
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: Maybe rename this flag to `is_initialized`?

File: docs/design/ftlib.md [hunk 17]
```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
...
+
+1. New workers report joining in
+2. Collective-communication operations fail or time out
+3. *Liveness probe fails for existing workers*
```
Comment (terrytangyuan) [inline docs/design/ftlib.md]: May need explain how the liveness probe is being done for the workers

File: docs/design/communication_library.md (hunk 3 above)
Comment (zw0610) [reply docs/design/communication_library.md]: touché

File: docs/design/consensus.md (hunk 6 above)
Comment (zw0610) [reply docs/design/consensus.md]: Could you elaborate it?

File: docs/design/ftlib.md (hunk 11 above)
Comment (zw0610) [reply docs/design/ftlib.md]: here I mean "wrapped high-level operations that executed when `FTLib.execute()` is called"

File: docs/design/ftlib.md (hunk 17 above)
Comment (zw0610) [reply docs/design/ftlib.md]: Sure, I added the following paragraph.

There are different approaches for liveness probe. *(For example, we force a `clean_ip` function for each worker at exit and scan the shared storage to check the existence of ip files as liveness probe. However, in the gossip-based consensus protocol, each worker will pull/push gossip messages of liveness from/to random nodes as liveness probe.)*

File: docs/design/consensus.md (hunk 8 above)
Comment (zw0610) [reply docs/design/consensus.md]: I changed "consenting member list" to "checking if member list has changed".

File: docs/design/communication_library.md (hunk 4 above)
Comment (zw0610) [reply docs/design/communication_library.md]: Yes, please check the updated version.

File: docs/design/consensus.md (hunk 7 above)
Comment (zw0610) [reply docs/design/consensus.md]: Here is how I describe the "rank-assign scheme":

We map the identification of each worker with a hash function and sort the workers with respect to hash value. The worker with minimal hash value will be considered as the *master* worker. While the *master* worker acts identically as other workers in the training phase, it may have additional components or operations during initialization or other stages.
